pkg/discovery/monitoring/kubelet/kubelet_monitor.go

```diff
@@ -0,0 +1,174 @@
+
+}
+
+func (m *KubeletMonitor) parseNodeInfo(node *api.Node, machineInfo *cadvisorapi.MachineInfo) {
+	cpuFrequencyMHz := float64(machineInfo.CpuFrequency) / util.MegaToKilo
+	cpuFrequencyMetric := metrics.NewEntityStateMetric(task.NodeType, util.NodeKeyFunc(node), metrics.CpuFrequency, cpuFrequencyMHz)
+	m.metricSink.AddNewMetricEntries(cpuFrequencyMetric)
+}
+
+// Parse node stats and put it into sink.
+func (m *KubeletMonitor) parseNodeStats(nodeStats stats.NodeStats) {
+	// cpu
+	cpuUsageCore := float64(*nodeStats.CPU.UsageNanoCores) / util.NanoToUnit
+	glog.V(4).Infof("Cpu usage of node %s is %f core", nodeStats.NodeName, cpuUsageCore)
+	nodeCpuUsageCoreMetrics := metrics.NewEntityResourceMetric(task.NodeType, util.NodeStatsKeyFunc(nodeStats),
+		metrics.CPU, metrics.Used, cpuUsageCore)
+
+	// memory
+	memoryUsageKiloBytes := float64(*nodeStats.Memory.UsageBytes) / util.KilobytesToBytes
+	glog.V(4).Infof("Memory usage of node %s is %f Kb", nodeStats.NodeName, memoryUsageKiloBytes)
+	nodeMemoryUsageKiloBytesMetrics := metrics.NewEntityResourceMetric(task.NodeType,
+		util.NodeStatsKeyFunc(nodeStats), metrics.Memory, metrics.Used, memoryUsageKiloBytes)
+
+	m.metricSink.AddNewMetricEntries(nodeCpuUsageCoreMetrics, nodeMemoryUsageKiloBytesMetrics)
+
+}
+
+// Parse pod stats for every pod and put them into sink.
+func (m *KubeletMonitor) parsePodStats(podStats []stats.PodStats) {
+	for _, podStat := range podStats {
+		var cpuUsageNanoCoreSum uint64
+		var memoryUsageBytesSum uint64
+		for _, containerStat := range podStat.Containers {
+			if containerStat.CPU != nil {
+				cpuUsageNanoCoreSum += *containerStat.CPU.UsageNanoCores
+			}
+			if containerStat.Memory != nil {
+				memoryUsageBytesSum += *containerStat.Memory.UsageBytes
+			}
+		}
+		glog.V(4).Infof("Cpu usage of pod %s is %f core", util.PodStatsKeyFunc(podStat),
+			float64(cpuUsageNanoCoreSum)/util.NanoToUnit)
+		podCpuUsageCoreMetrics := metrics.NewEntityResourceMetric(task.PodType, util.PodStatsKeyFunc(podStat),
+			metrics.CPU, metrics.Used, float64(cpuUsageNanoCoreSum)/util.NanoToUnit)
+
+		glog.V(4).Infof("Memory usage of pod %s is %f Kb", util.PodStatsKeyFunc(podStat),
+			float64(memoryUsageBytesSum)/util.KilobytesToBytes)
+		podMemoryUsageCoreMetrics := metrics.NewEntityResourceMetric(task.PodType, util.PodStatsKeyFunc(podStat),
+			metrics.Memory, metrics.Used, float64(memoryUsageBytesSum)/util.KilobytesToBytes)
+
+		// application cpu and mem used are the same as pod's.
+		applicationCpuUsageCoreMetrics := metrics.NewEntityResourceMetric(task.ApplicationType,
+			util.PodStatsKeyFunc(podStat), metrics.CPU, metrics.Used,
+			float64(cpuUsageNanoCoreSum)/util.NanoToUnit)
+		applicationMemoryUsageCoreMetrics := metrics.NewEntityResourceMetric(task.ApplicationType,
+			util.PodStatsKeyFunc(podStat), metrics.Memory, metrics.Used,
+			float64(memoryUsageBytesSum)/util.KilobytesToBytes)
+
+		m.metricSink.AddNewMetricEntries(podCpuUsageCoreMetrics,
```

Comment:
Performance concerns:
    accumulate the items into bulk, and add them to cache in one lock request.